### DISCUSSION TITLE
Editing Basic code with PC-Basic

### DISCUSSION THREAD

**Discussion by PAL-Thailand** [robhagemans/pcbasic#226] [Q&A]
I'm a newbie to GitHub and consequently may not be doing things correctly.  If that happens, please let me know.

My present concern is working with Rob Hagemans' PC-Basic app.  I don't seem to find the way to edit an existing BAS file. I have found that it IS possible to edit a specific line, but that's not what I'm hoping to do.

I would like to be able to move around in a specific BAS file to make improvements or add new code lines, etc.

Hopefully, my request might be resolved by the community or even by Mr. Hagemans (but I don't yet know how to make direct contact with him). 

I looked around the Internet for some enlightenment, but was not rewarded with a jackpot answer to my concern.

I have no idea what will happen once I press the "Start discussion" button below,  but here goes.....

**Comment by robhagemans** [comment]
Hi, editing files in the PC-BASIC interface is done by working with line numbers. To add a new line between two existing lines, choose a line number that is between them. There is no built-in text editor as this was not a feature of GW-BASIC, of which this is a close re-implementation. 

If you want to move around and edit a `.BAS` file, the best thing is to be sure to save it as text and use a text editor.

**Comment by Marrin** [comment]
And if there is no line number between existing numbers any more, remember that there is a command to `RENUM`ber existing lines, including fixing commands addressing the changing lines.

**Comment by PAL-Thailand** [comment]
Hello, dear Friends,

So, there is no full-fledged editor available.  Fortunately your suggestion  of saving the BAS file and using a text editor sounded OK. So, I went back to my ACER windows 7 laptop, fired up PC-Basic and made a simple single line change to the file and used the command 'save "filename.bas" ' either as shown or followed by the letter "A".  I was surprised to receive the message "Permission Denied." 

I prepared an attachment, but could not attach it to the post as suggested below.  I wonder why? Please see the attachment.

**Reply by Marrin** [reply]
Text from the PC-BASIC window can be copied to the clipboard by marking it with the mouse.

Where did you try to save the file?  Maybe it's just what it says: You (the current user) is not allowed to save anything there.  Or there is already a file with that name and it is write protected.  Or another software has opened the file you are trying to overwrite.

**Comment by PAL-Thailand** [comment]
Your reply to me suggested a variety of possibilities.
You asked "Where did you try to save the file? "As I was 
working within the PC-Basic main window, I had no idea 
where PC-Basic normally stores its files, but thought 
that it would be in the same folder where the appplication
was located. I thought this as I had experienced such an 
action from having use QBasic some time before.

Then you noted "Maybe it's just what it says: You (the 
current user) is not allowed to save anything there.". As 
the only user, could not understand why PC-Basic would not
logically put saved files in the same folder where 
everything else related to PC-Basic was stored. 

Then you continued by noting that "Or there is already a 
file with that name...". Well, that could and was the case
 since ALL PC-Basic folders and files were located.  
PC-Basic could have opted for suggesting to rename the 
file for saving -- but of course it opted to just say 
'Permission Denied'.  Not overly understandable, is it? 

You also pointed out that the file might be "write 
protected". And who would have write protected the file
in the first place?  I certainly did not write protect 
the file myself having no reason whatsoever to do so.

And finally, you wrote "Or another software has opened
 the file you are trying to overwrite."  That was not 
the case as the only program being used was PC-Basic!

So I am stil unable to move forward and need some detailed
advice for the following:

1) How to check if a file is 'write protected' or not?

2) If a file IS write protected, how can the write 
protection be removed?

3) Assuming that copying the open file with its newly 
edited line over to the clipboard for further transfer 
to an external text editor would be the proper way to 
save the edited .BAS file brings to mind the following 
concern of getting the text editor's final product back 
into PC-Basic's control?

These questions may have simple answers for the experts, 
but being a PC-Basic newbie leaves me quite in a dark room 
with little or no way of moving about.  Consequently, I 
would be very grateful for any help that long-time users 
of PC-Basic would be willing to share with me.

**Reply by Marrin** [reply]
QBasic uses the current working directory to store files.  That doesn't have to be the directory the qbasic.exe file is in.  The same should be true for PC-BASIC.  The standard should be drive Z: and that should be pointing to the current working directory when PC-BASIC was started.  You can list the contents of the current (virtual) drive and directory with the FILES command.  As far as I know there is no way from within PC-BASIC to get where the virtual drives are mounted/pointing to.

The PC-BASIC documentation has more information about [accessing your drives](http://robhagemans.github.io/pcbasic/doc/2.0/#mounting) under different operating systems.

Of course PC-BASIC can overwrite existing files.  But not existing *and* write protected files.

As to where such a protection might come from: We do know only what you reveal to us.  As it seems to be an existing file, it's reasonable to assume it was stored in an archive file of some kind (ZIP, LHA, ZOO, …) and some archives also store access rights (and/or ownership), which may lead to write protected files or directories after unpacking them.  Checking and changing file and directory access rights depends on the operating system.

Copying a program via clipboard to a file and back isn't really a good option unless the program is really short.  Anything longer than a screenful and it gets tedious and error prone.  There are LOAD, SAVE, and related commands to load and save BASIC programs and they generally work.

**Comment by PAL-Thailand** [comment]
Oops!  I forgot to ask if there is some way to edit a just created post in order to correct a typo?

**Reply by Marrin** [reply]
There should be three horizontal dots in the upper right corner of the posting that open a menu which should contain an „Edit“ entry.

**Comment by PAL-Thailand** [comment]
Hello, Marrin,

Thanks for your guidance.  Several points:  1) in this new post, I don't see the three dots in the upper right corner as I am typing.  However,  I DO see the three dots in that position on previous posts;  but clicking on the three dots in a previous post does not allow me to edit the post.

2) It would be most helpful to me to be able to provide you attachments showing clearly what I am seeing on my system as I work with PC-Basic.  But the notice at the bottom of a post, i.e. "Attach files by dragging & dropping, selecting or pasting them." is not responding to any attempt of mine to attach any file at all.  And it is not because I do not know how to drag & drop or select and paste.

Can we please solve these two above problems before returning to the "Permission Denied" problem?  As a PC-Pasic newbie, I am probably not pointing out various situations or details that most probably would help others in resolving the "Permission Denied".  And it is certainly true that a picture is worth a thousand words!

And now that I have finished my post, I still do not see any button that clearly tells me to 'post' the message to the forum.  So, I'm adding a third point to the about two above, making a total points that would help me greatly in communicating to the forum members.

**Reply by Marrin** [reply]
The three dots only appear in already posted messages. You wanted to edit a message — that's already possible when editing a message, so it doesn't make much sense to offer a menu with options that apply only to already published texts.

I get an „Edit“ option in that menu.  Of course only on my own comments and replies.  You can't edit messages of other people.

When I click on the „Attach files…“ line I get a file selection dialog with a filter on supported filename extensions preset.  Github doesn't allow just any file format as it seems.  On the other hand files with *.gz, *.tar, or *.zip can contain anything.  So it seems Github doesn't want to ban file formats, but ensure that things than can't be displayed in a browser or are known to be compressed anyway, are compressed before being attached.

But again: As long as it's just text it can be copied as text into the message.  For instance the banner printed at startup, marked with pressing the left mouse button and dragging over the text in the PC-BASIC window, and then pasting it with <kbd>Ctrl+V</kbd> into a fenced code block here:
```
PC-BASIC 2.0.6
(C) Copyright 2013--2022 Rob Hagemans.
60300 Bytes free
Ok 
```

Here's the somewhat shortest way to get a file I can load, but not save again, given an already existing `TEST.BAS` in the working directory → drive `Z:` in PC-BASIC.  First making the file read-only by revoking all _write_ permissions (this is under Linux):
```ShellSession
$ ls -l TEST.BAS
-rw-rw-r-- 1 bj bj 1063 Jul 16 19:41 TEST.BAS
$ chmod -w TEST.BAS
$ ls -l TEST.BAS
-r--r--r-- 1 bj bj 1063 Jul 16 19:41 TEST.BAS
```
Then in PC-BASIC loading and saving it:
```
load"test
Ok 
save"test
Permission Denied
```

So with relative short text snippets there is no need for attachments.

If you don't see a button to post a comment here, how _did_ you post that comment after all?  There's a „Comment“ button for sending new comments and a „Reply“ for sending when replying to a comment.
